### PR TITLE
Tickets/dm 35748

### PR DESCRIPTION
**Pull request by teweitsai**
* Add the **MockControlClosedLoop** class.
* Use the **TS_CONFIG_MTTCS_DIR** to get the configuration files.

### CODE REVIEW

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 1]
```diff
@@ -0,0 +1,860 @@
...
+        Returns
+        -------
+        force_balance : `dict`
+            Data of the force balance system.
+        """
+
+        axial_forces = self.axial_forces["hardpointCorrection"]
+        tangent_forces = self.tangent_forces["hardpointCorrection"]
+
+        net_forces = self._calc_net_forces_total(axial_forces, tangent_forces)
```
Comment (tribeiro) [inline python/lsst/ts/m2com/mock/mock_control_closed_loop.py:500]: I am not sure how widely you are using `self._calc_net_forces_total` and `self._calc_net_momments_total`, so take this comment with a grain of salt. 

But, consider what is happening when you call them like this. You are creating a variable with the data from a class member and then passing all that data to a function of the same class.. There's a lot of data allocation and transport going on here. 

Now, if you are only using it to compute things from class data that are in `self.axial_forces` and `self.tangent_forces`, you could potentially change the method to receive the dictionary key and use the data from the class directly.. For example:

```
def _calc_net_forces_total(self, force):
        angle = np.deg2rad(self._cell_geom["locAct_tangent"])

         fx = np.sum(np.cos(angle) * self.tangent_forces[force])
         fy = np.sum(np.sin(angle) * self.tangent_forces[force])
         fz = np.sum(self.axial_forces[force])

         return {"fx": fx, "fy": fy, "fz": fz}
```

This way there's no data allocation and transport happening, just local data access... 

But again, this only works if you are only using the method with the internal data.

**Review by tribeiro — APPROVED** [review]
The code looks good, I did left a couple comments I hope you will take a look before merging. 

Also, one thing that worries me a bit and you may want to take a second look is how many methods in `MockControlCloseLoop` receives and returns NumPy arrays. Some of the ones I looked at are (at least on a first glance) only using internal data from the class. Passing around big arrays in python like this can have a huge impact in efficiency.  NumPy arrays are very good for inline computation but passing them around to functions like this might be problematic.

Anyway, I'd take a second look at how you are implementing these methods and see if you can find ways to use the data "locally" instead of passing them around as function arguments. I made a suggestion inline I am not sure will work but, please, take a look.

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 2]
```diff
@@ -0,0 +1,860 @@
...
+    def reset_force_offsets(self):
+        """Reset the force offsets.
+
+        This will put the applied force to be zero.
+        """
+
+        self.axial_forces["applied"] = np.zeros(NUM_ACTUATOR - NUM_TANGENT_LINK)
+        self.tangent_forces["applied"] = np.zeros(NUM_TANGENT_LINK)
+
+    def get_net_forces_total(self):
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py:416]: This is to be consistent to the XML telemetry:
https://github.com/lsst-ts/ts_xml/blob/develop/sal_interfaces/MTM2/MTM2_Telemetry.xml#L358-L359

This was named by Bo. I think he tried to express the idea that this is for the "net" forces of "total" actuators.

We can update this at the next time when we update the xml interface. Thanks!

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 3]
```diff
@@ -0,0 +1,860 @@
...
+
+        angle = np.deg2rad(self._cell_geom["locAct_tangent"])
+
+        fx = np.sum(np.cos(angle) * tangent_forces)
+        fy = np.sum(np.sin(angle) * tangent_forces)
+        fz = np.sum(axial_forces)
+
+        return {"fx": fx, "fy": fy, "fz": fz}
+
+    def get_net_moments_total(self):
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py:452]: As above, this is to be consistent with the xml interface. But we could update the xml in a latter time. Thanks!

File: Jenkinsfile [hunk 4]
```diff
@@ -41,13 +45,29 @@ pipeline {
...
+        }
+
         stage('Unit Tests and Coverage Analysis') {
             steps {
                 // Pytest needs to export the junit report.
                 withEnv(["WORK_HOME=${env.WORKSPACE}"]) {
                     sh """
                         source ${env.SAL_SETUP_FILE}
 
+                        cd ${WORK_HOME}/ts_config_mttcs
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply Jenkinsfile]: Fixed in the commit: [a5cae23](https://github.com/lsst-ts/ts_m2com/pull/5/commits/a5cae233e833c27abcc68b23f3a62c505749c53f). Thanks!

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 5]
```diff
@@ -0,0 +1,860 @@
...
+        # Look-up table (LUT)
+        self._lut = dict()
+
+        # Cell geometry used in the calculation of net total forces and moments
+        self._cell_geom = dict()
+
+        # Hardpoint compensation matrix
+        self._hd_comp = np.array([])
+
+    def _get_default_temperatures(
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py]: Fixed in the commit: [a5cae23](https://github.com/lsst-ts/ts_m2com/pull/5/commits/a5cae233e833c27abcc68b23f3a62c505749c53f). Thanks!

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 6]
```diff
@@ -0,0 +1,860 @@
...
+        Raises
+        ------
+        `ValueError`
+            When the input dimensions of forces are wrong.
+        """
+
+        if (len(force_axial) != (NUM_ACTUATOR - NUM_TANGENT_LINK)) or (
+            len(force_tangent) != NUM_TANGENT_LINK
+        ):
+            raise ValueError("The input dimensions of forces are wrong.")
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py]: Fixed in the commit: [a5cae23](https://github.com/lsst-ts/ts_m2com/pull/5/commits/a5cae233e833c27abcc68b23f3a62c505749c53f). Thanks!

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 7]
```diff
@@ -0,0 +1,860 @@
...
+
+        self.axial_forces["measured"] = force_axial
+        self.tangent_forces["measured"] = force_tangent
+
+    def read_file_lut(self, filepath):
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py]: Fixed in the commit: [a5cae23](https://github.com/lsst-ts/ts_m2com/pull/5/commits/a5cae233e833c27abcc68b23f3a62c505749c53f). Thanks!

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 8]
```diff
@@ -0,0 +1,860 @@
...
+        `bool`
+            True if the cell temperature is high. Otherwise, False.
+        """
+
+        temp_exhaust = self.temperature["exhaust"]
+        temp_intake = self.temperature["intake"]
+
+        diff = (temp_exhaust[0] - temp_intake[0] + temp_exhaust[1] - temp_intake[1]) / 2
+
+        return True if diff > self.temperature["maxDiff"] else False
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py]: Fixed in the commit: [a5cae23](https://github.com/lsst-ts/ts_m2com/pull/5/commits/a5cae233e833c27abcc68b23f3a62c505749c53f). Thanks!

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 9]
```diff
@@ -0,0 +1,860 @@
...
+        -------
+        `dict`
+            Total net forces in Newton.
+        """
+        return self._calc_net_forces_total(
+            self.axial_forces["measured"], self.tangent_forces["measured"]
+        )
+
+    def _calc_net_forces_total(self, axial_forces, tangent_forces):
+        """Calculate the total net forces in Newton.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py]: Fixed in the commit: [a5cae23](https://github.com/lsst-ts/ts_m2com/pull/5/commits/a5cae233e833c27abcc68b23f3a62c505749c53f). Thanks!

File: python/lsst/ts/m2com/mock/mock_control_closed_loop.py [hunk 10]
```diff
@@ -0,0 +1,860 @@
...
+            ]
+        )
+        self.tangent_forces["lutTemperature"] = np.array([])
+
+        # These are hard-coded in the M2 cell LabVIEW project by vendor
+        self.hardpoints = [5, 15, 25, 73, 75, 77]
+
+        self.in_position_hardpoints = False
+
+        # Look-up table (LUT)
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2com/mock/mock_control_closed_loop.py]: Fixed in the commit: [a5cae23](https://github.com/lsst-ts/ts_m2com/pull/5/commits/a5cae233e833c27abcc68b23f3a62c505749c53f). Thanks!

**Comment by teweitsai** [comment]
For the concern of numpy array, I will take some time to do some test and bench mark to see its impact. Thanks for the suggestion!

**Pull request merged**